### ISSUE TITLE
[HOLD for payment 2024-01-24] Mweb-Compose-(*&~) not applied for multiline-quote message&edit compose of (_),displays extra lines

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#26941]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Open a report
3. Enter a multiline message, in that single quote(>) message include.
4. For message in above format, apply markdown ~, *&_.
5. Send the each type of markdown message seperately
5. Long press the message sent
6. Tap edit comment

## Expected Result:
When user applies markdown (~, *&_) for a multiline message with quote message, markdown must be applied correctly.
In edit compose box, all messages should be in the same format in which it was sent.

## Actual Result:
When user applies markdown (~ , *&_ ) for a multiline message with quote message, markdown not applied correctly for bold and strikethrough(*& ~ ), and correctly applied only for italic markdown.
In edit compose box, all messages should be in the same format in which it was sent.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.65-0

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856 

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/d48e3ae6-7953-4db1-b78a-95085305e453

https://github.com/Expensify/App/assets/115492554/d17e72af-3333-4670-ac2c-21bb9a3a4d87

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause-Internal Team

**Slack conversation:** @

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01236cf31874b5d2bb</li>
        <li>Upwork Job ID: 1701592423306964992</li>
        <li>Last Price Increase: 2023-09-19</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by johncschuster** [comment]
Triaged

**Comment by cubuspl42** [comment]
In this state, this issue contributes to the overall misunderstanding. It's the opposite, _inline_ syntax (like emphasis or strikethrough, `~`/`_`/`*`) effectively applied to a _block_ (in this case, a quote block) is an unexpected result.

**Comment by s77rt** [comment]
Can we put this on hold for https://github.com/Expensify/App/issues/27208? cc @johncschuster

**Comment by s77rt** [comment]
Not overdue. Let's put this on hold

**Comment by johncschuster** [comment]
Just put it on hold.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by s77rt** [comment]
This is on hold Melvin. @johncschuster should the title strictly have HOLD instead of HOLDING?

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
@parasharrajat This is about inconsistency in markdown rules, if you send a message that contains a quote and try to style the whole message it only works for certain styles:

Case 1: Message followed by a quote. All styled with strikethrough
```
~Msg then a
> quote~
```

Case 2: Message followed by a quote. All styled with bold
```
*Msg then a
> quote*
```

Case 2: Message followed by a quote. All styled with italic
```
_Msg then a
> quote_
```

You will notice that both case 1 and case 2 results in unstyled message (the literal `~` and `*` are printed instead) but for the italic message it works.

<img width="841" alt="Screenshot 2023-09-29 at 3 26 56 PM" src="https://github.com/Expensify/App/assets/16493223/a5be3267-39d1-4b0a-a8ca-6867bd8a17ec">

**Comment by cubuspl42** [comment]
@parasharrajat

> Correctly, I am not able to understand anything at all from the issue description.

Correctly -> Currently (?)

I shared my interpretation here:

https://github.com/Expensify/App/issues/26941#issuecomment-1717153957

[Like @s77rt noticed](https://github.com/Expensify/App/issues/26941#issuecomment-1740989845), we're dealing with inconsistency. The misunderstanding is that it is suggested that behavior labeled as "Case 1" and "Case 2" is the issue while actually the issue is "Case 3".

**Comment by cubuspl42** [comment]
@s77rt Case 3 is 100.00% a bug if ExpensiMark is anything like Markdown. The first `_` should belong to a paragraph block `_Mest then a` followed by a quote block `> quote`. Inlines (like `*` or `_`) are applied _within_ blocks in Markdown, not across them. The discussed syntax is Markdown-compatible, so going with a fundamentally different model sounds like a bad idea to me.

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
This can be `External` again and it's sill on hold

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
Same ^

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Bumping this to Weekly while we wait

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
https://github.com/Expensify/App/issues/27208 is now closed but didn't fix this issue. @johncschuster I think we can open this one (unless the other one got reopened)

**Comment by byte1012** [comment]
@johncschuster @s77rt 
Looks like there is no other opened one.
can i submit a proposal here ? or need to wait for remove 'holding'

**Comment by s77rt** [comment]
@byte1012 You can post a proposal here

**Comment by byte1012** [comment]
# Proposal
### Please re-state the problem that we are trying to solve in this issue.
When user applies markdown (~, *&_) for a multiline message with quote message, markdown must be applied correctly.

### What is the root cause of that problem?
Thoroughout the several platforms test (slack, github), we can check that expected result is like following
![image](https://github.com/Expensify/App/assets/86615752/dd108848-9a3d-4082-990f-63d1979be448)

Currently, Bold and Strikethrough markdown's behavior is correct, italic rule is not
It's because Italic rule is before blockquote rule and therefore Italic rule has been applied.

### What changes do you think we should make in order to solve the problem?
To fix it, we should move 'quote' rule before 'italic' rule
It can cause email inside of quote not to be parsed, to prevent that, we need to add a tweak on the autoemail rule's regex pattern

```
<previous>
                regex: new RegExp(
                    `([^\\w'#%+-]|^)${CONST.REG_EXP.MARKDOWN_EMAIL}(?!((?:(?!<a).)+)?<\\/a>|[^<>]*<\\/(?!em|h1))`,
                    'gim',
                ),
<after>
                regex: new RegExp(
                    `([^\\w'#%+-]|^)${CONST.REG_EXP.MARKDOWN_EMAIL}(?!((?:(?!<a).)+)?<\\/a>|[^<>]*<\\/(?!em|h1|blockquote))`,
                    'gim',
                ),

```

We can add the relevant test cases.
and above fixing does not make other test cases failed

**Comment by s77rt** [comment]
@byte1012 Thanks for the proposal. Your RCA is correct and the solution looks good to me :+1:

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/26941#issuecomment-1850051964)

**Comment by byte1012** [comment]
@s77rt @puneetlath 
Thanks for review and assignment
The PR for expensify-common https://github.com/Expensify/expensify-common/pull/618 is ready

**Comment by johncschuster** [comment]
I'm OOO starting Monday, December 18 (today) through Tuesday, January 2.

**Current status:**
PR has been submitted. We need to wait on the regression period to issue payment.

If this issue is open when I'm back from OOO, I'll take it back over. Thank you!

**Comment by byte1012** [comment]
@s77rt @puneetlath 
The PR for App https://github.com/Expensify/App/issues/26941 is ready

**Comment by greg-schroeder** [comment]
@byte1012 please follow the steps here: https://github.com/Expensify/App/issues/26941#issuecomment-1849902172

**Comment by byte1012** [comment]
Contributor details
Your Expensify account email: kelleyaquilino@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01f484b51cfedb480a

**Comment by byte1012** [comment]
Contributor details
Your Expensify account email: kelleyaquilino@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01f484b51cfedb480a

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified : https://github.com/Expensify/expensify-common/pull/566
- The offending PR has been commented on: https://github.com/Expensify/expensify-common/pull/566#discussion_r1457533789
- A discussion in #expensify-bugs has been started: Not needed
- Determine if we should create a regression test for this bug: No. We have added tests to `expensify-common` to prevent the bug from resurfacing again

**Comment by byte1012** [comment]
@puneetlath @johncschuster 
It looks like regression period is ended
Could you check payment please ?

**Comment by puneetlath** [comment]
@johncschuster is the BZ on this one and will take care of it. @johncschuster will you be able to pay it out today?

**Comment by byte1012** [comment]
@johncschuster could you send me offer on upwork ?

**Comment by byte1012** [comment]
@puneetlath @johncschuster 
Friendly bump !
@puneetlath could you do it instead ? Looks like @johncschuster is not available

**Comment by johncschuster** [comment]
Sorry I missed your pings, @byte1012! I'll take care of it right now.

**Comment by johncschuster** [comment]
@byte1012 @s77rt the offer has been sent! Can you accept that?

**Comment by s77rt** [comment]
@johncschuster Accepted! Thanks!

**Comment by byte1012** [comment]
@johncschuster friendly bump
can you please finish contract on upwork ? or do we wait for other ?

**Comment by byte1012** [comment]
@puneetlath could you please ping @johncschuster to check if he is available ?

**Comment by johncschuster** [comment]
Ah, sorry team. I've been struggling with keeping up with my notifications. I'm working on this now.

**Comment by johncschuster** [comment]
Payment has been issued! 🎉
